src/Microsoft.DncEng.Configuration.Extensions/LocalDevTokenCredential.cs

```diff
@@ -0,0 +1,124 @@
+                await File.WriteAllBytesAsync(CacheFilePath, protectedBytes);
+            }
+        }
+
+        private IAccount _account;
+
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            bool uiRequired = false;
+            if (_account != null)
+            {
+                try
+                {
+                    var result = await _app.AcquireTokenSilent(requestContext.Scopes, _account)
+                        .ExecuteAsync(cancellationToken);
+                    return new AccessToken(result.AccessToken, result.ExpiresOn);
+                }
+                catch (MsalUiRequiredException)
+                {
+                    // will be handled lower
+                    uiRequired = true;
+                }
+            }
+
+            // we give the lock a total wait time of 1 hour because the breakpoint to do device code authentication is inside the lock
+            using (await SentinelFileLock.AcquireAsync(DeviceCodeLockPath, 3600, TimeSpan.FromMilliseconds(1000)))
+            {
+                _account = Task.Run(async () => await _app.GetAccountsAsync(), cancellationToken).Result.FirstOrDefault();
+                if (_account == null || uiRequired)
+                {
+                    var deviceCodeResult = await _app.AcquireTokenWithDeviceCode(requestContext.Scopes, DeviceCodeResultCallback)
+                        .ExecuteAsync(cancellationToken);
+                    _account = deviceCodeResult.Account;
+                    return new AccessToken(deviceCodeResult.AccessToken, deviceCodeResult.ExpiresOn);
+                }
+            }
+            
+            // a recursive call here is fine because _account will not be null and the call will hit the first if statement
+            return await GetTokenAsync(requestContext, cancellationToken);
+        }
+
+        [MethodImpl(MethodImplOptions.NoOptimization)]
+        private static Task DeviceCodeResultCallback(DeviceCodeResult arg)
+        {
+            if (!Debugger.IsAttached)
+                throw new InvalidOperationException("Debugger required for local service fabric authentication.");
+
+            var userCode = arg.UserCode;
+            var verificationUrl = arg.VerificationUrl;
+            // If your debugger breaks here you need to authenticate to azure with a device code
+            // navigate a browser to the url contained in the "verificationUrl" variable and enter the code in the "userCode" variable
+            Debugger.Break();
+            return Task.CompletedTask;
+        }
+
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            return Task.Run(async () => await GetTokenAsync(requestContext, cancellationToken), cancellationToken).Result;
```

Comment:
`async () => await thing()` is just `() => thing()`. Why the extra async/await?

Reply:
because GetTokenAsync() returns `ValueTask<T>`. If I didn't have the `async () => await thing()` this Task.Run would be returning Task<ValueTask<T>>.